Add jsonschema validation for task options

chmeliik opened the pull request:
* OSBS-7672

The schema is used to validate arguments for the BuildContainerTask
handler() method, it validates the types of the individual arguments
(src: string, target: string, opts: dict or none) as well as the types
of parameters in the 'opts' dictionary. The schema for the 'opts' dict
is not complete, it allows unknown parameters because some options are
missing. This can be amended in the future.

chmeliik commented:
I have no idea why the fedora29+python2 combo fails, but it seems to fail when running `test.sh` from the `master` branch as well.

MartinBasti commented:
I'm curios if `builder_containerbuild.py` is better place for this check because `hub_containerbuild` just constructs a call which can be made without it anyway.

I'd place that check into `BuildContainerTask.handler`

MartinBasti commented:
For start we could check types for:
```
scratch=opts.get('scratch', False),
isolated=opts.get('isolated', False),
yum_repourls=opts.get('yum_repourls', None),
branch=opts.get('git_branch', None),
push_url=opts.get('push_url', None),
koji_parent_build=opts.get('koji_parent_build'),
release=release_overwrite,
flatpak=flatpak,
compose_ids=opts.get('compose_ids', None),
signing_intent=opts.get('signing_intent', None),
```

chmeliik commented:
Moved validation from hub_containerbuild to builder_containerbuild and added type checks for the values @MartinBasti suggested.

athos-ribeiro commented on koji_containerbuild/plugins/builder_containerbuild.py:
```diff
@@ -58,6 +60,10 @@
 
+# root of koji_containerbuild module
+TOP_DIR = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+BUILD_OPTS_SCHEMA_PATH = os.path.join(TOP_DIR, 'schemas', 'build_opts.json')
```
nitpick: should this file be packaged under a data directory instead?

chmeliik replied:
yeah that seems like a better place for it, done

chmeliik commented:
Now packaging the jsonschema definition file under /usr/share/

rcerven commented on koji-containerbuild.spec:
```diff
@@ -54,6 +54,11 @@ Summary:    Hub plugin that extend Koji to build layered container images
 Group:      Applications/System
 Requires:   koji-containerbuild
 Requires:   koji-hub
+%if 0%{with python3}
```
shouldn't this be under %package builder ?

chmeliik replied:
yes it should, fixed :)

twaugh commented on koji_containerbuild/plugins/builder_containerbuild.py:
```diff
@@ -58,6 +60,8 @@
 else:
     osbs_flatpak_support = True
 
+BUILD_PARAMS_SCHEMA_PATH = '/usr/share/koji-containerbuild/schemas/build_params.json'
```
Could use `pkg_resources.resource_stream` to get this path I think. See how atomic-reactor does it: https://github.com/containerbuildsystem/atomic-reactor/blob/1.6.41/atomic_reactor/util.py#L1269-L1274

chmeliik replied:
If I'm understanding it correctly, `pkg_resources.resource_stream` works in atomic-reactor because the json files are packaged under the `atomic_reactor` module (probably somewhere in `/usr/lib*/python*/atomic_reactor/`). Is there a way to use it for files packaged under `/usr/share/`? Or should the schema file for koji-containerbuild be packaged in the same way that atomic-reactor does it?

twaugh replied:
We can stick with /usr/share, no big deal.

athos-ribeiro commented on koji_containerbuild/plugins/builder_containerbuild.py:
```diff
@@ -58,6 +60,8 @@
 else:
     osbs_flatpak_support = True
 
+BUILD_PARAMS_SCHEMA_PATH = '/usr/share/koji-containerbuild/schemas/build_params.json'
```
Nice! this seems to be the proper place for the file.

Should the python package also install it here? I mean, what happens if this package is installed with `pip install`? You can take a look at how we do it in atomic-reactor for reference (check the `setup.py` file).

MartinBasti replied:
`/usr` part cannot be hardcoded, `pip install --user` or installation into virtualenv places `share` directory to somewhere else. This part mus be detected dynamically.

chmeliik commented:
Moved the installation of the jsonschema file from `koji-containerbuild.spec` to `setup.py`, thanks @athos-ribeiro (also squashed the related commits into one)

MartinBasti commented on schemas/build_params.json:
```diff
@@ -0,0 +1,66 @@
+        "signing_intent": {
+          "type": ["string", "null"],
+          "description": "Signing intent of the ODCS composes."
+        }
```
Don't we need here to allow extra properties?

chmeliik replied:
I think extra properties are allowed by default (there is a test for it, and it passes)

chmeliik commented:
After a decent amount of googling, I did not find any sane way to make `pkg_resources` tell you where it installed data_files. What we could do instead is install the json file under `/usr/lib/python*/koji_containerbuild/`, which is the same thing that we do in atomic-reactor. Will squash commit if approved.

athos-ribeiro commented:
> After a decent amount of googling, I did not find any sane way to make `pkg_resources` tell you where it installed data_files. What we could do instead is install the json file under `/usr/lib/python*/koji_containerbuild/`, which is the same thing that we do in atomic-reactor. Will squash commit if approved.
> 
> E: forgot to remove `%files` entry from specfile, fixed now

We could use fallbacks. We could, for example, look for the file under /usr/share, then look for it under the project path. This way we would both comply with [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) and support user installations.

chmeliik commented:
> We could use fallbacks. We could, for example, look for the file under /usr/share, then look for it under the project path. This way we would both comply with [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) and support user installations.

We could, but it would IMO be slightly more ugly than it is now, and keeping json files in python modules seems to be a fairly common practice (jsonschema itself does this, for example) so I wouldn't worry about it unless you think having the file in /usr/share is important.

chmeliik commented:
Squashed commits.

MartinBasti commented:
Release notes:
- input parameters are validated by schema and should provide nicer error reporting when parameters mismatch
